Build failure on Apple Silicon arm64 Macs

### How to reproduce ###

1. Build ccache 4.1 with Xcode 12.2 on macOS 11.0.1 on an Apple Silicon (arm64) Mac

### Actual behavior ###

Build failure. [Here is the full log](https://build.macports.org/builders/ports-11_arm64-builder/builds/3355/steps/install-port/logs/stdio). The errors are all of this type:

```
In file included from src/third_party/blake3/blake3_sse41.c:3:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include/immintrin.h:14:
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include/mmintrin.h:50:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
    return (__m64)__builtin_ia32_vec_init_v2si(__i, 0);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include/mmintrin.h:129:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
```
Despite the fact that the error messages appear to refer to files that are part of the development environment, I do not believe that there is a problem with the development environment.

### Expected behavior ###
<!-- What did you expect to happen? -->

Successful build.

### Environment ###
<!-- Which ccache version did you use? -->

ccache version 4.1
macOS 11.0.1
Xcode 12.2
Apple Silicon Mac

--- a/cmake/GenerateConfigurationFile.cmake
+++ b/cmake/GenerateConfigurationFile.cmake
@@ -77,7 +77,18 @@ if((CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_
   message(STATUS "Detected unsupported compiler for HAVE_AVX2 - disabled")
   set(HAVE_AVX2 FALSE)
 else()
-  check_cxx_compiler_flag(-mavx2 HAVE_AVX2)
+  set(CMAKE_REQUIRED_FLAGS "-mavx2")
+    check_c_source_compiles(
+      [=[
+        #include <immintrin.h>
+        int main()
+        {
+          _mm256_abs_epi8(_mm256_set1_epi32(42));
+          return 0;
+        }
+      ]=]
+      HAVE_AVX2)
+    unset(CMAKE_REQUIRED_FLAGS)
 endif()
 
 list(APPEND CMAKE_REQUIRED_LIBRARIES ws2_32)

--- a/cmake/GenerateConfigurationFile.cmake
+++ b/cmake/GenerateConfigurationFile.cmake
@@ -67,29 +67,19 @@ check_struct_has_member("struct stat" st_mtim sys/stat.h
 check_struct_has_member("struct statfs" f_fstypename sys/mount.h
                         HAVE_STRUCT_STATFS_F_FSTYPENAME)
 
-include(CheckCXXCompilerFlag)
-
-# Old GCC versions don't have the required header support.
-# Old Apple Clang versions seem to support -mavx2 but not the target
-# attribute that's used to enable AVX2 for a certain function.
-if((CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 5.0)
-   OR (CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 8.0))
-  message(STATUS "Detected unsupported compiler for HAVE_AVX2 - disabled")
-  set(HAVE_AVX2 FALSE)
-else()
-  set(CMAKE_REQUIRED_FLAGS "-mavx2")
-    check_c_source_compiles(
-      [=[
-        #include <immintrin.h>
-        int main()
-        {
-          _mm256_abs_epi8(_mm256_set1_epi32(42));
-          return 0;
-        }
-      ]=]
-      HAVE_AVX2)
-    unset(CMAKE_REQUIRED_FLAGS)
-endif()
+include(CheckCXXSourceCompiles)
+check_cxx_source_compiles(
+  [=[
+    #include <immintrin.h>
+    void func() __attribute__((target("avx2")));
+    void func() { _mm256_abs_epi8(_mm256_set1_epi32(42)); }
+    int main()
+    {
+      func();
+      return 0;
+    }
+  ]=]
+  HAVE_AVX2)
 
 list(APPEND CMAKE_REQUIRED_LIBRARIES ws2_32)
 list(REMOVE_ITEM CMAKE_REQUIRED_LIBRARIES ws2_32)

